File: core/amber/src/main/scala/edu/uci/ics/texera/web/resource/dashboard/file/UserFileResource.scala
package edu.uci.ics.texera.web.resource.dashboard.file

import com.google.common.io.Files
import edu.uci.ics.texera.web.SqlServer
import edu.uci.ics.texera.web.auth.SessionUser
import edu.uci.ics.texera.web.model.jooq.generated.Tables.{FILE, USER, USER_FILE_ACCESS}
import edu.uci.ics.texera.web.model.jooq.generated.tables.daos.{
  FileDao,
  FileOfProjectDao,
  UserDao,
  UserFileAccessDao
}
import edu.uci.ics.texera.web.model.jooq.generated.tables.pojos.{File, User}
import edu.uci.ics.texera.web.resource.dashboard.file.UserFileResource.{
  DashboardFileEntry,
  context,
  saveUserFileSafe
}
import io.dropwizard.auth.Auth
import org.apache.commons.lang3.tuple.Pair
import org.glassfish.jersey.media.multipart.{FormDataContentDisposition, FormDataParam}
import org.jooq.DSLContext
import org.jooq.types.UInteger

import java.io.{FileInputStream, IOException, InputStream, OutputStream}
import java.nio.file.Paths
import java.sql.Timestamp
import java.time.Instant
import java.util
import javax.annotation.security.PermitAll
import javax.ws.rs.core.{MediaType, Response, StreamingOutput}
import javax.ws.rs.{WebApplicationException, _}
import scala.collection.JavaConverters._
import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer
import java.net.URLDecoder

/**
  * Model `File` corresponds to `core/new-gui/src/app/common/type/user-file.ts` (frontend).
  */

object UserFileResource {
  private val context: DSLContext = SqlServer.createDSLContext
  private val fileDao = new FileDao(context.configuration)

  def saveUserFileSafe(
      uid: UInteger,
      fileName: String,
      uploadedInputStream: InputStream,
      size: UInteger,
      description: String
  ): String = {

    val fileNameStored = UserFileUtils.storeFileSafe(uploadedInputStream, fileName, uid)

    // insert record after completely storing the file on the file system.
    fileDao.insert(
      new File(
        uid,
        null,
        size,
        fileNameStored,
        UserFileUtils.getFilePath(uid, fileNameStored).toString,
        description,
        Timestamp.from(Instant.now())
      )
    )

    // insert UserFileAccess record to grant write access
    val fid = context
      .select(FILE.FID)
      .from(FILE)
      .where(FILE.UID.eq(uid).and(FILE.NAME.eq(fileNameStored)))
      .fetchOneInto(FILE)
      .getFid
    UserFileAccessResource.grantAccess(uid, fid, "write")
    fileNameStored
  }

  case class DashboardFileEntry(
      ownerName: String,
      accessLevel: String,
      isOwner: Boolean,
      file: File,
      projectIDs: List[UInteger]
  )
}

@PermitAll
@Path("/user/file")
@Consumes(Array(MediaType.APPLICATION_JSON))
@Produces(Array(MediaType.APPLICATION_JSON))
class UserFileResource {
  final private val fileDao = new FileDao(context.configuration)
  final private val userFileAccessDao = new UserFileAccessDao(
    context.configuration
  )
  final private val userDao = new UserDao(context.configuration)
  final private val fileOfProjectDao = new FileOfProjectDao(context.configuration)

  /**
    * This method will handle the request to upload a single file.
    * @return
    */
  @POST
  @Path("/upload")
  @Consumes(Array(MediaType.MULTIPART_FORM_DATA))
  def uploadFile(
      @FormDataParam("file") uploadedInputStream: InputStream,
      @FormDataParam("file") fileDetail: FormDataContentDisposition,
      @FormDataParam("size") size: UInteger,
      @FormDataParam("description") description: String,
      @Auth sessionUser: SessionUser
  ): Response = {
    val user = sessionUser.getUser
    val uid = user.getUid
    val fileName = fileDetail.getFileName
    val validationResult = validateFileName(fileName, uid)
    if (!validationResult.getLeft) {
      return Response
        .status(Response.Status.BAD_REQUEST)
        .entity(validationResult.getRight)
        .build()
    }
    saveUserFileSafe(uid, fileName, uploadedInputStream, size, description)
    Response.ok().build()
  }

  /**
    * This method returns a list of all files accessible by the current user
    *
    * @return
    */
  @GET
  @Path("/list")
  def listUserFiles(@Auth sessionUser: SessionUser): util.List[DashboardFileEntry] = {
    val user = sessionUser.getUser
    getUserFileRecord(user)
  }

  @GET
  @Path("/autocomplete/{query:.*}")
  def autocompleteUserFiles(
      @Auth sessionUser: SessionUser,
      @PathParam("query") q: String
  ): util.List[String] = {
    // get the user files
    // select the filenames that applies the input
    val query = URLDecoder.decode(q, "UTF-8")
    val user = sessionUser.getUser
    val fileList: List[DashboardFileEntry] = getUserFileRecord(user).asScala.toList
    val filenames = ArrayBuffer[String]()
    val username = user.getName
    // get all the filename list
    for (i <- fileList) {
      filenames += i.file.getName
    }
    // select the filenames that apply
    val selectedByFile = ArrayBuffer[String]()
    val selectedByUsername = ArrayBuffer[String]()
    val selectedByFullPath = ArrayBuffer[String]()
    for (e <- filenames) {
      val fullPath = username + "/" + e
      if (e.contains(query) || query.isEmpty)
        selectedByFile += (username + "/" + e)
      else if (username.contains(query))
        selectedByUsername += (username + "/" + e)
      else if (fullPath.contains(query))
        selectedByFullPath += (username + "/" + e)
    }
    (selectedByFile ++ selectedByUsername ++ selectedByFullPath).toList.asJava
  }

  private def getUserFileRecord(user: User): util.List[DashboardFileEntry] = {
    // fetch the user files:
    // user_file_access JOIN file on FID (to get all files this user can access)
    // then JOIN USER on UID (to get the name of the file owner)
    val fileRecords = context
      .select()
      .from(USER_FILE_ACCESS)
      .join(FILE)
      .on(USER_FILE_ACCESS.FID.eq(FILE.FID))
      .join(USER)
      .on(FILE.UID.eq(USER.UID))
      .where(USER_FILE_ACCESS.UID.eq(user.getUid))
      .fetch()

    // fetch the entire table of fileOfProject in memory, assuming this table is small
    val fileOfProjectMap = fileOfProjectDao
      .findAll()
      .asScala
      .groupBy(record => record.getFid)
      .mapValues(values => values.map(v => v.getPid).toList)

    val fileEntries: mutable.ArrayBuffer[DashboardFileEntry] = mutable.ArrayBuffer()
    fileRecords.forEach(fileRecord => {
      val file = fileRecord.into(FILE)
      val ownerUser = fileRecord.into(USER)
      val access = fileRecord.into(USER_FILE_ACCESS)

      var accessLevel = "None"
      if (access.getWriteAccess) {
        accessLevel = "Write"
      } else if (access.getReadAccess) {
        accessLevel = "Read"
      } else {
        accessLevel = "None"
      }
      val ownerName = ownerUser.getName
      val projectIDs = fileOfProjectMap.getOrElse(file.getFid, List())
      fileEntries += DashboardFileEntry(
        ownerName,
        accessLevel,
        ownerName == user.getName,
        new File(file),
        projectIDs
      )
    })
    fileEntries.toList.asJava
  }

  /**
    * This method deletes a file from a user's repository
    * @param fileName the name of file being deleted
    * @param ownerName the name of the file's owner
    * @return
    */
  @DELETE
  @Path("/delete/{fileName}/{ownerName}")
  def deleteUserFile(
      @PathParam("fileName") fileName: String,
      @PathParam("ownerName") ownerName: String,
      @Auth sessionUser: SessionUser
  ): Response = {

    val user = sessionUser.getUser
    val fileID = UserFileAccessResource.getFileId(ownerName, fileName)
    val userID = user.getUid
    val hasWriteAccess = context
      .select(USER_FILE_ACCESS.WRITE_ACCESS)
      .from(USER_FILE_ACCESS)
      .where(USER_FILE_ACCESS.UID.eq(userID).and(USER_FILE_ACCESS.FID.eq(fileID)))
      .fetch()
      .getValue(0, 0)
    if (hasWriteAccess == false) {
      Response
        .status(Response.Status.UNAUTHORIZED)
        .entity("You do not have the access to deleting the file")
        .build()
    } else {
      val filePath = fileDao.fetchOneByFid(fileID).getPath
      UserFileUtils.deleteFile(Paths.get(filePath))
      fileDao.deleteById(fileID)
      Response.ok().build()
    }
  }

  @POST
  @Path("/validate")
  @Consumes(Array(MediaType.MULTIPART_FORM_DATA))
  def validateUserFile(
      @FormDataParam("name") fileName: String,
      @Auth sessionUser: SessionUser
  ): Response = {
    val user = sessionUser.getUser
    val validationResult = validateFileName(fileName, user.getUid)
    if (validationResult.getLeft)
      Response.ok().build()
    else {
      Response.status(Response.Status.BAD_REQUEST).entity(validationResult.getRight).build()
    }

  }

  private def validateFileName(fileName: String, userID: UInteger): Pair[Boolean, String] = {
    if (fileName == null) Pair.of(false, "file name cannot be null")
    else if (fileName.trim.isEmpty) Pair.of(false, "file name cannot be empty")
    else if (isFileNameExisted(fileName, userID)) Pair.of(false, "file name already exists")
    else Pair.of(true, "filename validation success")
  }

  private def isFileNameExisted(fileName: String, userID: UInteger): Boolean =
    context.fetchExists(
      context
        .selectFrom(FILE)
        .where(FILE.UID.equal(userID).and(FILE.NAME.equal(fileName)))
    )

  @GET
  @Path("/download/{fileId}")
  def downloadFile(
      @PathParam("fileId") fileId: UInteger,
      @Auth sessionUser: SessionUser
  ): Response = {
    val user = sessionUser.getUser
    val filePath: Option[java.nio.file.Path] =
      UserFileUtils.getFilePathByIds(user.getUid, fileId)
    if (filePath.isDefined) {
      val fileObject = filePath.get.toFile

      // sending a FileOutputStream/ByteArrayOutputStream directly will cause MessageBodyWriter
      // not found issue for jersey
      // so we create our own stream.
      val fileStream = new StreamingOutput() {
        @throws[IOException]
        @throws[WebApplicationException]
        def write(output: OutputStream): Unit = {
          val data = Files.toByteArray(fileObject)
          output.write(data)
          output.flush()
        }
      }
      Response
        .ok(fileStream, MediaType.APPLICATION_OCTET_STREAM)
        .header(
          "content-disposition",
          String.format("attachment; filename=%s", fileObject.getName)
        )
        .build
    } else {

      Response
        .status(Response.Status.BAD_REQUEST)
        .`type`(MediaType.TEXT_PLAIN)
        .entity(s"Could not find file $fileId of ${user.getName}")
        .build()
    }

  }

  /**
    * This method updates the name of a given userFile
    *
    * @param file the to be updated file
    * @return the updated userFile
    */
  @POST
  @Path("/update/name")
  @Consumes(Array(MediaType.APPLICATION_JSON))
  @Produces(Array(MediaType.APPLICATION_JSON))
  def changeUserFileName(file: File, @Auth sessionUser: SessionUser): Unit = {
    val userId = sessionUser.getUser.getUid
    val fid = file.getFid
    val newFileName = file.getName

    val validationRes = this.validateFileName(newFileName, userId)
    val hasWriteAccess = context
      .select(USER_FILE_ACCESS.WRITE_ACCESS)
      .from(USER_FILE_ACCESS)
      .where(USER_FILE_ACCESS.UID.eq(userId).and(USER_FILE_ACCESS.FID.eq(fid)))
      .fetch()
      .getValue(0, 0)
    if (hasWriteAccess == false) {
      throw new ForbiddenException("No sufficient access privilege.")
    }
    if (validationRes.getLeft == false) {
      throw new BadRequestException(validationRes.getRight)
    } else {
      val userFile = fileDao.fetchOneByFid(fid)
      val filePath = userFile.getPath

      val uploadedInputStream = new FileInputStream(filePath)
      // delete the original file
      UserFileUtils.deleteFile(Paths.get(filePath))
      // store the file with the new file name
      val fileNameStored = UserFileUtils.storeFileSafe(uploadedInputStream, newFileName, userId)

      userFile.setName(newFileName)
      userFile.setPath(UserFileUtils.getFilePath(userId, fileNameStored).toString)
      fileDao.update(userFile)
    }
  }

  /**
    * This method updates the description of a given userFile
    *
    * @param file the to be updated file
    * @return the updated userFile
    */
  @POST
  @Path("/update/description")
  @Consumes(Array(MediaType.APPLICATION_JSON))
  @Produces(Array(MediaType.APPLICATION_JSON))
  def changeUserFileDescription(file: File, @Auth sessionUser: SessionUser): Unit = {
    val userId = sessionUser.getUser.getUid
    val fid = file.getFid
    val newFileDescription = file.getDescription

    val hasWriteAccess = context
      .select(USER_FILE_ACCESS.WRITE_ACCESS)
      .from(USER_FILE_ACCESS)
      .where(USER_FILE_ACCESS.UID.eq(userId).and(USER_FILE_ACCESS.FID.eq(fid)))
      .fetch()
      .getValue(0, 0)
    if (hasWriteAccess == false) {
      throw new ForbiddenException("No sufficient access privilege.")
    }
    val userFile = fileDao.fetchOneByFid(fid)
    userFile.setDescription(newFileDescription)
    fileDao.update(userFile)
  }
}

File: core/archive/old-engine/web/src/main/java/edu/uci/ics/texera/web/resource/UserFileResource.java
package edu.uci.ics.texera.web.resource;

import edu.uci.ics.texera.dataflow.resource.file.FileManager;
import edu.uci.ics.texera.dataflow.sqlServerInfo.SqlServer;
import edu.uci.ics.texera.web.TexeraWebException;
import edu.uci.ics.texera.web.response.GenericWebResponse;
import io.dropwizard.jersey.sessions.Session;
import org.apache.commons.lang3.tuple.Pair;
import org.glassfish.jersey.media.multipart.FormDataContentDisposition;
import org.glassfish.jersey.media.multipart.FormDataParam;
import org.jooq.Record1;
import org.jooq.Record5;
import org.jooq.Result;
import org.jooq.types.UInteger;

import javax.servlet.http.HttpSession;
import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import java.io.InputStream;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import static edu.uci.ics.texera.dataflow.jooq.generated.Tables.FILE;
import static org.jooq.impl.DSL.defaultValue;

@Path("/user/file")
@Consumes(MediaType.APPLICATION_JSON)
@Produces(MediaType.APPLICATION_JSON)
public class UserFileResource {

    /**
     * Corresponds to `src/app/common/type/user-file.ts`
     */
    public static class UserFile {
        public UInteger id; // the ID in MySQL database is unsigned int
        public String name;
        public String path;
        public String description;
        public UInteger size; // the size in MySQL database is unsigned int

        public UserFile(UInteger id, String name, String path, String description, UInteger size) {
            this.id = id;
            this.name = name;
            this.path = path;
            this.description = description;
            this.size = size;
        }
    }

    /**
     * This method will handle the request to upload a single file.
     *
     */
    @POST
    @Path("/upload")
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    public GenericWebResponse uploadFile(
            @FormDataParam("file") InputStream uploadedInputStream,
            @FormDataParam("file") FormDataContentDisposition fileDetail,
            @FormDataParam("size") String size,
            @FormDataParam("description") String description,
            @Session HttpSession session
    ) {
        UInteger userID = UserResource.getUser(session).getUserID();
        String fileName = fileDetail.getFileName();
        UInteger sizeUInteger = parseStringToUInteger(size);

        Pair<Boolean, String> validationResult = validateFileName(fileName, userID);
        if (!validationResult.getLeft()) {
            return new GenericWebResponse(1, validationResult.getRight());
        }

        this.handleFileUpload(uploadedInputStream, fileName, description, sizeUInteger, userID);
        return GenericWebResponse.generateSuccessResponse();
    }

    @GET
    @Path("/list")
    public List<UserFile> listUserFiles(@Session HttpSession session) {
        UInteger userID = UserResource.getUser(session).getUserID();

        Result<Record5<UInteger, String, String, String, UInteger>> result = getUserFileRecord(userID);

        if (result == null) return new ArrayList<>();

        List<UserFile> fileList = result.stream()
                .map(
                        record -> new UserFile(
                                record.get(FILE.FID),
                                record.get(FILE.NAME),
                                record.get(FILE.PATH),
                                record.get(FILE.DESCRIPTION),
                                record.get(FILE.SIZE)
                        )
                ).collect(Collectors.toList());

        return fileList;
    }

    @DELETE
    @Path("/delete/{fileID}")
    public GenericWebResponse deleteUserFile(@PathParam("fileID") String fileID, @Session HttpSession session) {
        UInteger userID = UserResource.getUser(session).getUserID();
        UInteger fileIdUInteger = parseStringToUInteger(fileID);
        Record1<String> result = deleteInDatabase(fileIdUInteger, userID);

        if (result == null) return new GenericWebResponse(1, "The file does not exist");

        String filePath = result.get(FILE.PATH);
        FileManager.getInstance().deleteFile(Paths.get(filePath));

        return GenericWebResponse.generateSuccessResponse();
    }

    @POST
    @Path("/validate")
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    public GenericWebResponse validateUserFile(@Session HttpSession session, @FormDataParam("name") String fileName) {
        UInteger userID = UserResource.getUser(session).getUserID();
        Pair<Boolean, String> validationResult = validateFileName(fileName, userID);
        return new GenericWebResponse(validationResult.getLeft() ? 0 : 1, validationResult.getRight());
    }

    private Record1<String> deleteInDatabase(UInteger fileID, UInteger userID) {
        /**
         * Known problem for jooq 3.x
         * delete...returning clause does not work properly
         * retrieve the filepath first, then delete it.
         */
        Record1<String> result = SqlServer.createDSLContext()
                .select(FILE.PATH)
                .from(FILE)
                .where(FILE.FID.eq(fileID).and(FILE.UID.equal(userID)))
                .fetchOne();

        int count = SqlServer.createDSLContext()
                .delete(FILE)
                .where(FILE.FID.eq(fileID).and(FILE.UID.equal(userID)))
                //.returning(USERFILE.FILEPATH) does not work
                .execute();

        throwErrorWhenNotOne("delete file " + fileID + " failed in database", count);

        return result;
    }

    private Result<Record5<UInteger, String, String, String, UInteger>> getUserFileRecord(UInteger userID) {
        return SqlServer.createDSLContext()
                .select(FILE.FID, FILE.NAME, FILE.PATH, FILE.DESCRIPTION, FILE.SIZE)
                .from(FILE)
                .where(FILE.UID.equal(userID))
                .fetch();
    }

    private void handleFileUpload(InputStream fileStream, String fileName, String description, UInteger size, UInteger userID) {
        int count = insertFileToDataBase(
                fileName,
                FileManager.getFilePath(userID.toString(), fileName).toString(),
                size,
                description,
                userID);

        throwErrorWhenNotOne("Error occurred while inserting file record to database", count);

        FileManager.getInstance().storeFile(fileStream, fileName, userID.toString());
    }

    private UInteger parseStringToUInteger(String userID) throws TexeraWebException {
        try {
            return UInteger.valueOf(userID);
        } catch (NumberFormatException e) {
            throw new TexeraWebException("Incorrect String to Double");
        }
    }


    private int insertFileToDataBase(String fileName, String path, UInteger size, String description, UInteger userID) {
        return SqlServer.createDSLContext().insertInto(FILE)
                .set(FILE.UID, userID)
                .set(FILE.FID, defaultValue(FILE.FID))
                .set(FILE.NAME, fileName)
                .set(FILE.PATH, path)
                .set(FILE.DESCRIPTION, description)
                .set(FILE.SIZE, size)
                .execute();
    }

    private Pair<Boolean, String> validateFileName(String fileName, UInteger userID) {
        if (fileName == null) {
            return Pair.of(false, "file name cannot be null");
        } else if (fileName.trim().isEmpty()) {
            return Pair.of(false, "file name cannot be empty");
        } else if (isFileNameExisted(fileName, userID)) {
            return Pair.of(false, "file name already exists");
        } else {
            return Pair.of(true, "filename validation success");
        }
    }

    private Boolean isFileNameExisted(String fileName, UInteger userID) {
        return SqlServer.createDSLContext()
                .fetchExists(
                        SqlServer.createDSLContext()
                                .selectFrom(FILE)
                                .where(FILE.UID.equal(userID)
                                        .and(FILE.NAME.equal(fileName)))
                );
    }

    /**
     * Most the sql operation should only be executed once. eg. insertion, deletion.
     * this method will raise TexeraWebException when the input number is not one
     *
     * @param errorMessage the message displaying when raising the error
     * @param number       the number to be checked with one
     * @throws TexeraWebException
     */
    private void throwErrorWhenNotOne(String errorMessage, int number) throws TexeraWebException {
        if (number != 1) {
            throw new TexeraWebException(errorMessage);
        }
    }
}
